### ISSUE TITLE
linuxkit build cannot encode header

### ISSUE THREAD

**Issue by Gabology** [linuxkit/linuxkit#3091]
**Description**
Running `linuxkit build` with the `docker.elastic.co/logstash/logstash-oss` repository fails.

**Steps to reproduce the issue:**
YAML file
```yml
kernel:
  image: linuxkit/kernel:4.14.52
  cmdline: "console=tty0 console=ttyS0 console=ttyAMA0"
init:
  - linuxkit/init:0e4af96fecc8f752c80d41c0b7d06570cc1dc6b2
  - linuxkit/runc:v0.4
  - linuxkit/containerd:27a4c84cc8fab2d5ea04342546ecd20453ec99b3
  - linuxkit/ca-certificates:v0.4
onboot:
  - name: sysctl
    image: linuxkit/sysctl:v0.4
  - name: dhcpcd
    image: linuxkit/dhcpcd:v0.4
    command: ["/sbin/dhcpcd", "--nobackground", "-f", "/dhcpcd.conf", "-1"]
  - name: metadata
    image: linuxkit/metadata:v0.4
services:
  - name: getty
    image: linuxkit/getty:44730fd0a7c59dbacf5b48b54ba33f551bcf7ef0
    env:
     - INSECURE=true
  - name: rngd
    image: linuxkit/rngd:v0.4
  - name: logstash
    image: "docker.elastic.co/logstash/logstash-oss:6.3.0"
    binds:
      - /etc/resolv.conf:/etc/resolv.conf
      - /run/config/logstash/config:/usr/share/logstash/config
      - /run/config/logstash/pipeline:/usr/share/logstash/pipeline
trust:
  org:
    - linuxkit
    - library
```

`linuxkit build -format iso-efi logstash.yml`

**Describe the results you received:**
```
Extract kernel image: linuxkit/kernel:4.14.52
Add init containers:
Process init image: linuxkit/init:0e4af96fecc8f752c80d41c0b7d06570cc1dc6b2
Process init image: linuxkit/runc:v0.4
Process init image: linuxkit/containerd:27a4c84cc8fab2d5ea04342546ecd20453ec99b3
Process init image: linuxkit/ca-certificates:v0.4
Add onboot containers:
  Create OCI config for linuxkit/sysctl:v0.4
  Create OCI config for linuxkit/dhcpcd:v0.4
  Create OCI config for linuxkit/metadata:v0.4
Add service containers:
  Create OCI config for linuxkit/getty:44730fd0a7c59dbacf5b48b54ba33f551bcf7ef0
  Create OCI config for linuxkit/rngd:v0.4
  Create OCI config for docker.elastic.co/logstash/logstash-oss:6.3.0
FATA[0022] Failed to extract root filesystem for docker.elastic.co/logstash/logstash-oss:6.3.0: archive/tar: cannot encode header: Format specifies USTAR; and USTAR cannot encode Linkname="containers/services/logstash/lower/usr/lib/python2.7/site-packages/kitchen/collections/strictdict.pyc"
```

**Describe the results you expected:**
That the image would build successfully.

**Additional information you deem important (e.g. issue happens only occasionally):**
OSX 10.13.5, docker version 18.03.1-ce

**Comment by Gabology** [comment]
Looks like it could possibly be related to this? https://github.com/golang/go/issues/24821

**Comment by Gabology** [comment]
https://github.com/moby/tool/pull/216 will fix this issue, feel free to close it when it has been integrated upstream.

**Comment by Gabology** [comment]
Thanks for the quick response! 

For my use case there isn't any hurry as I can build the image using the upstream `moby` tool until the fix has been integrated in linuxkit.

**Comment by masterzen** [comment]
I've recompiled latest `linuxkit` HEAD (26ae9a0246d4acbcd09d6b8e0d00896f3eed6925), which contains moby/tool#216 fix (note that I got the exact same issue with `moby/tool`).
This is with all tools compiled with go 1.10.3.

It turns out moby/tool#216 is not enough to fix the problem for my case (the culprit image is docker.elastic.co/elasticsearch/elasticsearch:5.6.10).

The reason is that `moby/tool` is blindly copying files from the incoming tar image to the outgoing tar image without changing the header format of the copied files. Since we "move" the files to a new prefix during that operation, the filename (or linkname) can become larger than the USTAR limit. If the original image was USTAR encoded (which is the case for the elasticsearch image), then you'll get a fatal error.

I'm proposing a very small fix in moby/tool#219. It would be great if this PR is accepted and merged to have it vendored again in `linuxkit`.

Thanks!
